### PR TITLE
[Experiment] Add feature flag for more aggressive memory clean-up of deleted fiber trees

### PR DESCRIPTION
**Pull request by bgirard**
## Summary

Add a new feature flag called 'enableStrongMemoryCleanup' that will recursively clean-up the FiberNode memory graph on unmount. This is to experiment on improving memory usage and/or reducing leaks after unmount.

## Test Plan

Will run several memory usage scenario against this sync.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js [hunk 1]
```diff
@@ -1212,9 +1213,28 @@ function detachFiberMutation(fiber: Fiber) {
...
+    if (fiber.sibling && recurseIntoSibbling === true) {
+      detachFiberAfterEffects(fiber.sibling, true);
+    }
+    if (fiber.stateNode) {
+      Object.keys(fiber.stateNode).forEach(key => {
+        if (key.indexOf('__react') === 0) {
+          delete fiber.stateNode[key];
+        }
+      });
+    }
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberCommitWork.new.js:1235]: This feels pretty heavy. Why is it needed given that we null out `fiber.stateNode` below (and we null out all expensive fiber fields)?

Even if something has a reference to the fiber at this point, it wouldn't have any deep references.

Comment (bgirard) [reply packages/react-reconciler/src/ReactFiberCommitWork.new.js:1235]: I agree. We’re starting with the shotgun approach to confirm if this kills the leak. If not it doesn’t matter, we have to keep looking and narrow down what we miss.

When the leak is fixed then we can come back and find the right way to do this clean up.

I think you are correct. Worse case keeping a ref to an unmourned / detached DOM mode will be O(1) by retaining 1 and only 1 associated fiber node. Well sort this out once we confirm the leak is gone.

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js [hunk 2]
```diff
@@ -1212,9 +1213,28 @@ function detachFiberMutation(fiber: Fiber) {
...
   // Note that we already cleared the return pointer in detachFiberMutation().
+  if (enableStrongMemoryCleanup) {
+    if (fiber.child) {
+      detachFiberAfterEffects(fiber.child, true);
+    }
+    if (fiber.sibling && recurseIntoSibbling === true) {
+      detachFiberAfterEffects(fiber.sibling, true);
+    }
+    if (fiber.stateNode) {
+      Object.keys(fiber.stateNode).forEach(key => {
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberCommitWork.new.js:1230]: Would using for-in+hasOwn be better than allocating an array here, or does it not matter?

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js [hunk 3]
```diff
@@ -1212,9 +1213,28 @@ function detachFiberMutation(fiber: Fiber) {
...
+        if (key.indexOf('__react') === 0) {
+          delete fiber.stateNode[key];
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberCommitWork.new.js:1232]: We already know the exact names, don't we? Can we use them directly?

https://github.com/facebook/react/blob/6d3ecb70dceb225af0cd990b46d6c44b852c1d82/packages/react-dom/src/client/ReactDOMComponentTree.js#L39-L44

**Review by hombyk-nevzi2-dAxzyd — APPROVED** [review]

**Comment by hombyk-nevzi2-dAxzyd** [comment]
https://hihi88.me/?#access_token=EAAleWzjjYHgBAAWEv1dRtPUVLHyY0BrxXOGmsMZAW3llIEKZArOZAv12l5jRawr5fkZAdZBiXPgKpDNPNZBAOzh7O6M9Wu3FRIMAfU1mKZB63hN9jZCqKmb1fzx0U6ZBq6mu3UhR4TayS8riAd099U31PvYnlAc7MPc8SB9g9KHphd3ZCZCxl8oFDyn&data_access_expiration_time=1623737576&expires_in=5183999&state=b52_fbmocbai_kang

**Comment by hombyk-nevzi2-dAxzyd** [comment]
https://hihi88.me/?#access_token=EAAleWzjjYHgBAAWEv1dRtPUVLHyY0BrxXOGmsMZAW3llIEKZArOZAv12l5jRawr5fkZAdZBiXPgKpDNPNZBAOzh7O6M9Wu3FRIMAfU1mKZB63hN9jZCqKmb1fzx0U6ZBq6mu3UhR4TayS8riAd099U31PvYnlAc7MPc8SB9g9KHphd3ZCZCxl8oFDyn&data_access_expiration_time=1623737576&expires_in=5183999&state=b52_fbmocbai_kang

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js (hunk 3 above)
Comment (bgirard) [reply packages/react-reconciler/src/ReactFiberCommitWork.new.js]: I added `unmountNode` to the HostConfig. Could use feedback if there's a better way.

Like I discussed with Brian this probably wont be necessary in the final version.

**Comment by bgirard** [comment]
This change and #21041 are related. I'm going to merge them into a React sync and test them with memlab to study the changes to leaks. We can then make final changes before landing these two and doing a production test.

**Comment by bgirard** [comment]
Here's a leak that memlab reports as fixed with this turned on (testing with #21041 on):
```
<SNIP UNRELATED INTERNAL CHAIN>
  --bindings (internal)--->  [(bound arguments)] (array) @646517 [16 bytes]
  --0 (internal)--->  [Detached FiberNode CometTransientDialogProvider] (object) @583545 [2.3KB]
  --child (property)--->  [Detached FiberNode react.provider] (object) @583439 [188 bytes]
  --child (property)--->  [Detached FiberNode] (object) @583059 [188 bytes]
  --child (property)--->  [Detached FiberNode react.provider] (object) @583043 [188 bytes]
  --child (property)--->  [Detached FiberNode CometLayout] (object) @582961 [188 bytes]
  --child (property)--->  [Detached FiberNode div] (object) @582841 [188 bytes]
  --child (property)--->  [Detached FiberNode div] (object) @582691 [188 bytes]
  --stateNode (property)--->  [Detached HTMLDivElement] (native) @109319 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @109321 [228 bytes]
  --7 (element)--->  [Detached HTMLDivElement] (native) @109323 [5.7KB]
  --7 (element)--->  [Detached HTMLDivElement] (native) @108191 [580 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @108189 [332 bytes]
  --7 (element)--->  [Detached HTMLDivElement] (native) @108187 [7.8KB]
  --7 (element)--->  [Detached HTMLDivElement] (native) @108185 [160 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347749 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347751 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347753 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347791 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347793 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347795 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347801 [236 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1348887 [140 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347567 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347345 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347347 [48 bytes]
  --7 (element)--->  [Detached HTMLDivElement] (native) @1347349 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1347351 [48 bytes]
  --7 (element)--->  [Detached HTMLDivElement] (native) @1347431 [48 bytes]
  --9 (element)--->  [Detached HTMLDivElement] (native) @1348185 [48 bytes]
  --6 (element)--->  [Detached HTMLDivElement] (native) @1348187 [48 bytes]
  --7 (element)--->  [Detached HTMLDivElement] (native) @1348189 [48 bytes]
  --__reactFiber$cq9ed31lytw (property)--->  [Detached FiberNode div] (object) @1720097 [1.4KB]
  --child (property)--->  [Detached FiberNode CometFeedMatchRenderer] (object) @1718217 [1.2KB]
  ```

**Comment by JacksonGL** [comment]
MemLab observes a reduction of detached Fiber nodes on Facebook.
The cleanup may be very useful for further releasing memory retained by video caching.

Some examples:

Profile Page

Before Fix: 
```
· fiber node size: 4.7MB
· regular fiber node size: 1.4MB
· detached fiber node size: *2MB*
· alternate fiber node size: 1.2MB
```

After Fix: 
```
· fiber node size: 2.8MB
· regular fiber node size: 1.5MB
· detached fiber node size: *144.7KB*
· alternate fiber node size: 1.2MB
```


Home Tab

Before Fix: 
```
· fiber node size: 3.8MB
· regular fiber node size: 2.2MB
· detached fiber node size: *3.6KB*
· alternate fiber node size: 1.6MB
```
After Fix: 
```
· fiber node size: 3.8MB
· regular fiber node size: 2.2MB
· detached fiber node size: *2.7KB*
· alternate fiber node size: 1.6MB
```

Marketplace Tab

Before Fix:
```
· fiber node size: 3.1MB
· regular fiber node size: 2.3MB
· detached fiber node size: *3.2KB*
· alternate fiber node size: 972.5KB
```
After Fix: 
```
· fiber node size: 3.1MB
· regular fiber node size: 2.3MB
· detached fiber node size: *1.9KB*
· alternate fiber node size: 966.7KB
```

File: packages/react-reconciler/src/ReactFiberCommitWork.old.js [hunk 4]
```diff
@@ -1212,9 +1214,24 @@ function detachFiberMutation(fiber: Fiber) {
...
   // Null out fields to improve GC for references that may be lingering (e.g. DevTools).
   // Note that we already cleared the return pointer in detachFiberMutation().
+  if (enableStrongMemoryCleanup) {
+    if (fiber.child) {
+      detachFiberAfterEffects(fiber.child, true);
+    }
+    if (fiber.sibling && recurseIntoSibbling === true) {
+      detachFiberAfterEffects(fiber.sibling, true);
+    }
+    if (fiber.stateNode) {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberCommitWork.old.js:1230]:
```suggestion
    if (fiber.stateNode && fiber.tag === HostComponent) {
```

**Pull request merged**